lib/ai/tools/shell/session/output.ts

```
@@ -0,0 +1,107 @@
+ * the canonical line with the full prompt (e.g. "user host % command").
+ */
+function deduplicateCommandLines(lines: string[], command: string): string[] {
+  const trimmedCmd = command.trim();
+  if (!trimmedCmd) return lines;
+
+  const escaped = escapeForRegex(trimmedCmd);
+  const bareRe = new RegExp(`^\\s*${escaped}\\s*$`);
+  const prefixRe = new RegExp(`^\\s*[$#]\\s*${escaped}\\s*$`);
+
+  const hasCanonicalLine = lines.some((line) => {
+    const s = line.replace(/\x1b\[[0-9;]*m/g, "").trim();
```

Biome (2.3.14)

[error] Unexpected control character in a regular expression.

Control characters are unusual and potentially incorrect inputs, so they are disallowed.

(lint/suspicious/noControlCharactersInRegex)